synapse/handlers/sso.py

```diff
@@ -693,8 +697,49 @@ async def _register_mapped_user(
         await self._store.record_user_external_id(
             auth_provider_id, remote_user_id, registered_user_id
         )
+
+        # Set avatar, if available
+        if attributes.picture:
+            await self.set_avatar(registered_user_id, attributes.picture)
+
         return registered_user_id
 
+    async def set_avatar(self, user_id: str, picture_https_url: str) -> None:
+        try:
+            uid = UserID.from_string(user_id)
+
+            # download picture
+            http_response = requests.get(picture_https_url)
+            if http_response.status_code != 200:
+                http_response.raise_for_status()
+
+            content_type = http_response.headers["Content-Type"]
+            content_length = int(http_response.headers["Content-Length"])
+
+            # convert image into BytesIO
+            b = io.BytesIO(http_response.content)
+
+            # store it in media repository
+            avatar_mxc_url = await self.media_repo.create_content(
```

Comment:
We should use the API instead of calling synapse internal code here, since the media repository can be outside of synapse, cf https://github.com/turt2live/matrix-media-repo.

Reply:
Following a talk on #synapse-dev, some more infos:
- we should use the public API and public facing endpoint, no internal routing should be needed
- we need to authenticate the POST `/upload`. For that I think we can inject a temporary access token in synapse, do the upload with that, and then delete this temporary token.

Reply:
> we should use the public API and public facing endpoint, no internal routing should be needed

Hrm, tricky. On the one hand, we'll have to use the public API for integration with matrix-media-repo. On the other hand:

 * This isn't really consistent with our architecture, and I don't think it will work reliably: in some deployments it may require connection hairpinning which may not work.  Having most inter-worker traffic go via internal APIs, and some go via public APIs, is unpleasant, architecturally speaking.
 * Using the public API means we need to implement authentication kludges like the one you suggest.

I think I agree that using the public API is the least bad, but as @dklimpel says, please make sure this new requirement is clearly documented.

(Also, if the media repo is in-process, we should go direct.)

Whatever happens, please put it in a separate "add this media to the media repo" function, rather than embedding it all in the SSO handler.

Reply:
> For that I think we can inject a temporary access token in synapse, do the upload with that, and then delete this temporary token.

Ugh. That sounds like a great way for us to end up with unused access tokens.

> I think I agree that using the public API is the least bad

No, I'm changing my mind here. We need a hybrid approach:
 * If the media repo is in-process, do it directly
 * If the media repo is a separate synapse worker, use an internal HTTP replication API (like we have with event persisters)
 * If the media repo is a separate matrix-media-repo, use its configured api with a configured [shared secret](https://docs.t2bot.io/matrix-media-repo/shared-secret/index.html)

We'll need to carefully consider what the configuration settings look like to make sure it is consistent with the other settings, and isn't limited to the SSO usecase.

Reply:
Maybe for now we just do the "in-process" solution (ie, not support it on worker-based deployments) and then extend it to the other mechanisms in later PRs?

Reply:
@richvdh Would you say this "in-process" solution is equivalent to how its currently being done via `self.media_repo.create_content()` call? I am not sure what should I try to do here.

Reply:
> Would you say this "in-process" solution is equivalent to how its currently being done via self.media_repo.create_content() call?

Yes. We need to:

- document that this isn't supported in worker deployments, or with external media repos,
- ideally not call `set_avatar` in the first place if we are not on the main process, or if we're using an external media repo.

The first bullet is the most important. I'm not completely sure how to do the second bullet off the top of my head.